File: output_20260319_145930/claude-opus/c/Struct/account.c
#include "stdlib.h"

struct account
{
    int balance;
};

/*@ predicate account(struct account *a;) =
    a->balance |-> _;
@*/

struct account *create_account()
//@ requires true;
//@ ensures result != 0 &*& account(result);
{
    struct account *myAccount = malloc(sizeof(struct account));
    if (myAccount == 0)
    {
        abort();
    }
    myAccount->balance = 0;
    //@ close account(myAccount);
    return myAccount;
}

void account_set_balance(struct account *myAccount, int newBalance)
//@ requires account(myAccount);
//@ ensures account(myAccount);
{
    //@ open account(myAccount);
    myAccount->balance = newBalance;
    //@ close account(myAccount);
}

void account_dispose(struct account *myAccount)
//@ requires account(myAccount);
//@ ensures true;
{
    //@ open account(myAccount);
    free(myAccount);
}

int main()
//@ requires true;
//@ ensures true;
{
    struct account *myAccount = create_account();
    account_set_balance(myAccount, 5);
    account_dispose(myAccount);
    return 0;
}

File: output_20260319_145930/claude-opus/rust/Struct/account.rs
use std::alloc::{Layout, alloc, handle_alloc_error, dealloc};

struct Account {
    balance: i32,
}

/*@
pred Account_own(p: *mut Account, bal: i32) =
    alloc_block(p as *mut u8, Layout::new_::<Account>()) &*&
    struct_Account_padding(p) &*&
    (*p).balance |-> bal;
@*/

impl Account {

    unsafe fn create() -> *mut Account
    //@ req true;
    //@ ens Account_own(result, 0);
    {
        let my_account = alloc(Layout::new::<Account>()) as *mut Account;
        if my_account.is_null() {
            handle_alloc_error(Layout::new::<Account>());
        }
        //@ close_struct(my_account);
        (*my_account).balance = 0;
        //@ close Account_own(my_account, 0);
        my_account
    }

    unsafe fn set_balance(my_account: *mut Account, new_balance: i32)
    //@ req Account_own(my_account, _);
    //@ ens Account_own(my_account, new_balance);
    {
        //@ open Account_own(my_account, _);
        (*my_account).balance = new_balance;
        //@ close Account_own(my_account, new_balance);
    }

    unsafe fn dispose(my_account: *mut Account)
    //@ req Account_own(my_account, _);
    //@ ens true;
    {
        //@ open Account_own(my_account, _);
        //@ open_struct(my_account);
        dealloc(my_account as *mut u8, Layout::new::<Account>());
    }

}

fn main()
//@ req true;
//@ ens true;
{
    unsafe {
        let my_account = Account::create();
        Account::set_balance(my_account, 5);
        Account::dispose(my_account);
    }
}